File: backend/app/models/checklist.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, func, Boolean, Text
from sqlalchemy.orm import relationship
from ..db.session import Base


class TuvTermin(Base):
    __tablename__ = "tuv_termine"

    id = Column(Integer, primary_key=True, index=True)
    fahrzeug_id = Column(Integer, ForeignKey("fahrzeuge.id"), nullable=False)
    ablauf_datum = Column(DateTime, nullable=False)
    status = Column(String(50), default="reminder")
    letzte_pruefung = Column(DateTime, nullable=True)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    fahrzeug = relationship("Fahrzeug", back_populates="tuv_termine")


class Checkliste(Base):
    __tablename__ = "checklisten"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    fahrzeuggruppe_id = Column(Integer, ForeignKey("fahrzeuggruppen.id"), nullable=False)
    ersteller_id = Column(Integer, ForeignKey("benutzer.id"), nullable=True)
    template = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    fahrzeuggruppe = relationship("FahrzeugGruppe", back_populates="checklisten")
    ersteller = relationship("Benutzer", back_populates="erstellte_checklisten")
    items = relationship("ChecklistItem", back_populates="checkliste", cascade="all, delete-orphan")
    ausfuehrungen = relationship("ChecklistAusfuehrung", back_populates="checkliste", cascade="all, delete-orphan")


class ChecklistItem(Base):
    __tablename__ = "checklist_items"

    id = Column(Integer, primary_key=True, index=True)
    checkliste_id = Column(Integer, ForeignKey("checklisten.id"), nullable=False)
    beschreibung = Column(String(500), nullable=False)
    pflicht = Column(Boolean, default=True)
    reihenfolge = Column(Integer, default=0)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    checkliste = relationship("Checkliste", back_populates="items")
    ergebnisse = relationship("ItemErgebnis", back_populates="item", cascade="all, delete-orphan")


class ChecklistAusfuehrung(Base):
    __tablename__ = "checklist_ausfuehrungen"

    id = Column(Integer, primary_key=True, index=True)
    checkliste_id = Column(Integer, ForeignKey("checklisten.id"), nullable=False)
    fahrzeug_id = Column(Integer, ForeignKey("fahrzeuge.id"), nullable=False)
    benutzer_id = Column(Integer, ForeignKey("benutzer.id"), nullable=False)
    status = Column(String(50), default="started")
    started_at = Column(DateTime, server_default=func.now(), nullable=False)
    completed_at = Column(DateTime, nullable=True)

    # Relationships
    checkliste = relationship("Checkliste", back_populates="ausfuehrungen")
    fahrzeug = relationship("Fahrzeug", back_populates="ausfuehrungen")
    benutzer = relationship("Benutzer", back_populates="ausfuehrungen")
    ergebnisse = relationship("ItemErgebnis", back_populates="ausfuehrung", cascade="all, delete-orphan")


class ItemErgebnis(Base):
    __tablename__ = "item_ergebnisse"

    id = Column(Integer, primary_key=True, index=True)
    ausfuehrung_id = Column(Integer, ForeignKey("checklist_ausfuehrungen.id"), nullable=False)
    item_id = Column(Integer, ForeignKey("checklist_items.id"), nullable=False)
    status = Column(String(50), default="ok")
    kommentar = Column(Text, nullable=True)
    created_at = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    ausfuehrung = relationship("ChecklistAusfuehrung", back_populates="ergebnisse")
    item = relationship("ChecklistItem", back_populates="ergebnisse")


class AuditLog(Base):
    __tablename__ = "audit_log"

    id = Column(Integer, primary_key=True, index=True)
    benutzer_id = Column(Integer, ForeignKey("benutzer.id"), nullable=True)
    aktion = Column(String(200), nullable=False)
    ressource_typ = Column(String(100), nullable=False)
    ressource_id = Column(Integer, nullable=True)
    alte_werte = Column(Text, nullable=True)
    neue_werte = Column(Text, nullable=True)
    timestamp = Column(DateTime, server_default=func.now(), nullable=False)

    # Relationships
    benutzer = relationship("Benutzer", back_populates="audit_logs")

File: android-webview-legacy/app/src/main/java/com/feuerwehr/checklist/data/models/Checklist.kt
package com.feuerwehr.checklist.data.models

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName
import kotlinx.datetime.Instant

/**
 * Checklist data models for the Fire Department Checklist App
 * Corresponds to backend Checkliste, ChecklistItem, ChecklistAusfuehrung, and ItemErgebnis models
 */

// Network DTOs for API communication
data class ChecklistDto(
    val id: Int,
    val name: String,
    @SerializedName("fahrzeuggruppe_id")
    val fahrzeuggruppeId: Int,
    @SerializedName("ersteller_id")
    val erstellerId: Int,
    val template: Boolean = false,
    @SerializedName("created_at")
    val createdAt: String,
    val items: List<ChecklistItemDto>? = null,
    val fahrzeuggruppe: VehicleGroupDto? = null
)

data class ChecklistItemDto(
    val id: Int,
    @SerializedName("checkliste_id")
    val checklisteId: Int,
    val beschreibung: String,
    val pflicht: Boolean = false, // Required item
    val reihenfolge: Int = 0, // Order in checklist
    @SerializedName("created_at")
    val createdAt: String
)

data class ChecklistAusfuehrungDto(
    val id: Int,
    @SerializedName("checkliste_id")
    val checklisteId: Int,
    @SerializedName("fahrzeug_id")
    val fahrzeugId: Int,
    @SerializedName("benutzer_id")
    val benutzerId: Int,
    val status: String, // "in_progress", "completed", "cancelled"
    @SerializedName("started_at")
    val startedAt: String,
    @SerializedName("completed_at")
    val completedAt: String?,
    val checkliste: ChecklistDto? = null,
    val fahrzeug: VehicleDto? = null,
    val ergebnisse: List<ItemErgebnisDto>? = null
)

data class ItemErgebnisDto(
    val id: Int,
    @SerializedName("ausfuehrung_id")
    val ausfuehrungId: Int,
    @SerializedName("item_id")
    val itemId: Int,
    val status: String, // "ok", "fehler", "nicht_pruefbar"
    val kommentar: String?,
    @SerializedName("created_at")
    val createdAt: String,
    val item: ChecklistItemDto? = null
)

// Request/Response DTOs
data class CreateChecklistRequest(
    val name: String,
    @SerializedName("fahrzeuggruppe_id")
    val fahrzeuggruppeId: Int,
    val template: Boolean = false,
    val items: List<CreateChecklistItemRequest>
)

data class CreateChecklistItemRequest(
    val beschreibung: String,
    val pflicht: Boolean = false,
    val reihenfolge: Int = 0
)

data class StartChecklistRequest(
    @SerializedName("checkliste_id")
    val checklisteId: Int,
    @SerializedName("fahrzeug_id")
    val fahrzeugId: Int
)

data class UpdateItemErgebnisRequest(
    val status: String,
    val kommentar: String?
)

// Room database entities
@Entity(tableName = "checklists")
data class ChecklistEntity(
    @PrimaryKey
    val id: Int,
    val name: String,
    val fahrzeuggruppeId: Int,
    val erstellerId: Int,
    val template: Boolean = false,
    val createdAt: String,
    val lastSyncedAt: Long = 0L,
    val isLocalOnly: Boolean = false
)

@Entity(tableName = "checklist_items")
data class ChecklistItemEntity(
    @PrimaryKey
    val id: Int,
    val checklisteId: Int,
    val beschreibung: String,
    val pflicht: Boolean = false,
    val reihenfolge: Int = 0,
    val createdAt: String,
    val lastSyncedAt: Long = 0L,
    val isLocalOnly: Boolean = false
)

@Entity(tableName = "checklist_ausfuehrungen")
data class ChecklistAusfuehrungEntity(
    @PrimaryKey
    val id: Int,
    val checklisteId: Int,
    val fahrzeugId: Int,
    val benutzerId: Int,
    val status: String,
    val startedAt: String,
    val completedAt: String?,
    val lastSyncedAt: Long = 0L,
    val isLocalOnly: Boolean = false
)

@Entity(tableName = "item_ergebnisse")
data class ItemErgebnisEntity(
    @PrimaryKey
    val id: Int,
    val ausfuehrungId: Int,
    val itemId: Int,
    val status: String,
    val kommentar: String?,
    val createdAt: String,
    val lastSyncedAt: Long = 0L,
    val isLocalOnly: Boolean = false
)

// Domain models for UI layer
data class Checklist(
    val id: Int,
    val name: String,
    val fahrzeuggruppeId: Int,
    val erstellerId: Int,
    val template: Boolean = false,
    val createdAt: Instant,
    val items: List<ChecklistItem> = emptyList(),
    val fahrzeuggruppe: VehicleGroup? = null,
    val isLocalOnly: Boolean = false
)

data class ChecklistItem(
    val id: Int,
    val checklisteId: Int,
    val beschreibung: String,
    val pflicht: Boolean = false,
    val reihenfolge: Int = 0,
    val createdAt: Instant,
    val isLocalOnly: Boolean = false
)

data class ChecklistAusfuehrung(
    val id: Int,
    val checklisteId: Int,
    val fahrzeugId: Int,
    val benutzerId: Int,
    val status: ChecklistStatus,
    val startedAt: Instant,
    val completedAt: Instant?,
    val checkliste: Checklist? = null,
    val fahrzeug: Vehicle? = null,
    val ergebnisse: List<ItemErgebnis> = emptyList(),
    val isLocalOnly: Boolean = false
) {
    val progress: Float
        get() = if (checkliste?.items?.isEmpty() != false) 0f
                else ergebnisse.size.toFloat() / checkliste.items.size.toFloat()

    val isComplete: Boolean
        get() = status == ChecklistStatus.COMPLETED

    fun getResultForItem(itemId: Int): ItemErgebnis? {
        return ergebnisse.find { it.itemId == itemId }
    }
}

data class ItemErgebnis(
    val id: Int,
    val ausfuehrungId: Int,
    val itemId: Int,
    val status: ItemStatus,
    val kommentar: String?,
    val createdAt: Instant,
    val item: ChecklistItem? = null,
    val isLocalOnly: Boolean = false
)

enum class ChecklistStatus(val value: String, val displayName: String) {
    IN_PROGRESS("in_progress", "In Bearbeitung"),
    COMPLETED("completed", "Abgeschlossen"),
    CANCELLED("cancelled", "Abgebrochen");

    companion object {
        fun fromString(status: String): ChecklistStatus {
            return entries.find { it.value == status } ?: IN_PROGRESS
        }
    }
}

enum class ItemStatus(val value: String, val displayName: String, val isError: Boolean = false) {
    OK("ok", "OK", false),
    FEHLER("fehler", "Fehler", true),
    NICHT_PRUEFBAR("nicht_pruefbar", "Nicht prüfbar", false);

    companion object {
        fun fromString(status: String): ItemStatus {
            return entries.find { it.value == status } ?: OK
        }
    }
}